examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp

```diff
@@ -120,123 +201,112 @@ std::array<int, 6> RegularGridDecomposition::getExtentOfSubdomain(const int subd
+  const std::array<double, _dimensionCount> globalBoxMax = {_globalBoxMax[0], _globalBoxMax[1], _globalBoxMax[2]};
+  const std::array<double, _dimensionCount> globalBoxLength =
+      autopas::utils::ArrayMath::sub(globalBoxMax, globalBoxMin);
+
+  for (int dimensionIndex = 0; dimensionIndex < _dimensionCount; ++dimensionIndex) {
+    std::vector<ParticleType> immigrants, remainingEmigrants;
+    std::vector<ParticleType> particlesForLeftNeighbour;
+    std::vector<ParticleType> particlesForRightNeighbour;
+
+    // See documentation for _neighbourDomainIndices to explain the indexing
+    int leftNeighbour = _neighbourDomainIndices[(dimensionIndex * 2) % _neighbourCount];
+    int rightNeighbour = _neighbourDomainIndices[(dimensionIndex * 2 + 1) % _neighbourCount];
+
+    for (const auto &particle : emigrants) {
+      std::array<double, _dimensionCount> position = particle.getR();
+      if (position[dimensionIndex] < _localBoxMin[dimensionIndex]) {
+        particlesForLeftNeighbour.push_back(particle);
+
+        // Apply boundary condition
+        if (_localBoxMin[dimensionIndex] == _globalBoxMin[dimensionIndex]) {
+          position[dimensionIndex] =
+              std::min(std::nextafter(_globalBoxMax[dimensionIndex], _globalBoxMin[dimensionIndex]),
+                       position[dimensionIndex] + globalBoxLength[dimensionIndex]);
+          particlesForLeftNeighbour.back().setR(position);
         }
-        emigrants = remainingEmigrants;
-
-        sendAndReceiveParticlesLeftAndRight(particlesForLeftNeighbour, particlesForRightNeighbour, leftNeighbour,
-                                            rightNeighbour, immigrants);
-
-        for (const auto &particle : immigrants) {
-          if (isInsideLocalDomain(particle.getR())) {
-            autoPasContainer->addParticle(particle);
-          } else {
-            emigrants.push_back(particle);
-          }
+      } else if (position[dimensionIndex] >= _localBoxMax[dimensionIndex]) {
+        particlesForRightNeighbour.push_back(particle);
+
+        // Apply boundary condition
+        if (_localBoxMax[dimensionIndex] == _globalBoxMax[dimensionIndex]) {
+          position[dimensionIndex] =
+              std::max(_globalBoxMin[dimensionIndex], position[dimensionIndex] - globalBoxLength[dimensionIndex]);
+          particlesForRightNeighbour.back().setR(position);
         }
+      } else {
+        remainingEmigrants.push_back(particle);
+      }
+    }
+    emigrants = remainingEmigrants;
+
+    sendAndReceiveParticlesLeftAndRight(particlesForLeftNeighbour, particlesForRightNeighbour, leftNeighbour,
+                                        rightNeighbour, immigrants);
 
-        immigrants.clear();
+    for (const auto &particle : immigrants) {
+      if (isInsideLocalDomain(particle.getR())) {
+        autoPasContainer->addParticle(particle);
+      } else {
+        emigrants.push_back(particle);
```

Comment:
you could think about calling `emigrants.reserve(emigrants.size() + immigrants.size())` or something like that. Maybe also only some fraction of `immigrants`?
Similar for the other `push_back()` loops.

Reply:
Does the assignment `emigrants = remainingEmigrants` change the amount of reserved space for the emigrants vector?
Because remainingEmigrants should be smaller than emigrants, there should be some unused space left for the incoming immigrants.